presto-delta/src/main/java/com/facebook/presto/delta/DeltaPageSourceProvider.java

```diff
@@ -0,0 +1,402 @@
+                typesBuilder.add(type);
+
+                if (isPushedDownSubfield(column)) {
+                    Subfield pushedDownSubfield = getPushedDownSubfield(column);
+                    List<String> nestedColumnPath = nestedColumnPath(pushedDownSubfield);
+                    Optional<ColumnIO> columnIO = findNestedColumnIO(lookupColumnByName(messageColumnIO, pushedDownSubfield.getRootName()), nestedColumnPath);
+                    if (columnIO.isPresent()) {
+                        fieldsBuilder.add(constructField(type, columnIO.get()));
+                    }
+                    else {
+                        fieldsBuilder.add(Optional.empty());
+                    }
+                }
+                else if (getParquetType(type, fileSchema, column, tableName, path).isPresent()) {
+                    String columnName = name;
+                    fieldsBuilder.add(constructField(type, lookupColumnByName(messageColumnIO, columnName)));
+                }
+                else {
+                    fieldsBuilder.add(Optional.empty());
+                }
+            }
+            return new ParquetPageSource(parquetReader, typesBuilder.build(), fieldsBuilder.build(), namesBuilder.build(), new RuntimeStats());
+        }
+        catch (Exception e) {
+            try {
+                if (dataSource != null) {
+                    dataSource.close();
+                }
+            }
+            catch (IOException ignored) {
+            }
+            if (e instanceof PrestoException) {
+                throw (PrestoException) e;
+            }
+            if (e instanceof ParquetCorruptionException) {
+                throw new PrestoException(DELTA_BAD_DATA, e);
+            }
+            if (e instanceof AccessControlException) {
+                throw new PrestoException(PERMISSION_DENIED, e.getMessage(), e);
+            }
+            if (nullToEmpty(e.getMessage()).trim().equals("Filesystem closed") || e instanceof FileNotFoundException) {
+                throw new PrestoException(DELTA_CANNOT_OPEN_SPLIT, e);
+            }
+            String message = format("Error opening Hive split %s (offset=%s, length=%s): %s", path, start, length, e.getMessage());
+            if (e.getClass().getSimpleName().equals("BlockMissingException")) {
+                throw new PrestoException(DELTA_MISSING_DATA, message, e);
+            }
+            throw new PrestoException(DELTA_CANNOT_OPEN_SPLIT, message, e);
+        }
+    }
+
+    public static TupleDomain<ColumnDescriptor> getParquetTupleDomain(Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<DeltaColumnHandle> effectivePredicate)
+    {
+        if (effectivePredicate.isNone()) {
+            return TupleDomain.none();
+        }
+
+        ImmutableMap.Builder<ColumnDescriptor, Domain> predicate = ImmutableMap.builder();
+        for (Map.Entry<DeltaColumnHandle, Domain> entry : effectivePredicate.getDomains().get().entrySet()) {
```

Comment:
`effectivePredicate.getDomains().get()` is called without a isPresent() check. Please add the validation.